llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2348 @@
+    dbgs() << "\n[addTemplateParams] DINodeArray\n";
+    // TParams->dump();
+    for (const auto *Entry : TParams)
+      Entry->dump();
+  });
+
+  // Add template parameters.
+  for (const auto *Entry : TParams) {
+    if (const auto *TTP = dyn_cast<DITemplateTypeParameter>(Entry))
+      constructTemplateTypeParameter(Element, TTP);
+    else if (const auto *TVP = dyn_cast<DITemplateValueParameter>(Entry))
+      constructTemplateValueParameter(Element, TVP);
+  }
+}
+
+// DISubprogram
+void LVIRReader::applySubprogramAttributes(LVScope *Function,
+                                           const DISubprogram *SP,
+                                           bool SkipSPAttributes) {
+  assert(Function && "Invalid logical element");
+  assert(SP && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[applySubprogramAttributes] DISubprogram\n";
+    SP->dump();
+  });
+
+  // If -fdebug-info-for-profiling is enabled, need to emit the subprogram
+  // and its source location.
+  bool SkipSPSourceLocation =
+      SkipSPAttributes && !getCUNode()->getDebugInfoForProfiling();
+  if (!SkipSPSourceLocation)
+    if (applySubprogramDefinitionAttributes(Function, SP, SkipSPAttributes))
+      return;
+
+  if (!SkipSPSourceLocation)
+    addSourceLine(Function, SP);
+
+  // Skip the rest of the attributes under -gmlt to save space.
+  if (SkipSPAttributes)
+    return;
+
+  DITypeRefArray Args;
+  if (const DISubroutineType *SPTy = SP->getType())
+    Args = SPTy->getTypeArray();
+
+  // Construct subprogram return type.
+  if (Args.size()) {
+    LVElement *ElementType = getOrCreateType(Args[0]);
+    Function->setType(ElementType);
+  }
+
+  // Add virtuality info if available.
+  Function->setVirtualityCode(SP->getVirtuality());
+
+  if (!SP->isDefinition()) {
+    // Add arguments. Do not add arguments for subprogram definition. They will
+    // be handled while processing variables.
+    constructSubprogramArguments(Function, Args);
+  }
+
+  if (SP->isArtificial())
+    Function->setIsArtificial();
+
+  if (!SP->isLocalToUnit())
+    Function->setIsExternal();
+
+  // Add accessibility info if available.
+  addAccess(Function, SP->getFlags());
+}
+
+// DISubprogram
+bool LVIRReader::applySubprogramDefinitionAttributes(LVScope *Function,
+                                                     const DISubprogram *SP,
+                                                     bool Minimal) {
+  assert(Function && "Invalid logical element");
+  assert(SP && "Invalid metadata node.");
+  LLVM_DEBUG({
+    dbgs() << "\n[applySubprogramDefinitionAttributes] DISubprogram\n";
+    SP->dump();
+  });
+
+  LVScope *Reference = nullptr;
+  StringRef DeclLinkageName;
+  if (const DISubprogram *SPDecl = SP->getDeclaration()) {
+    if (!Minimal) {
+      DITypeRefArray DeclArgs, DefinitionArgs;
+      DeclArgs = SPDecl->getType()->getTypeArray();
+      DefinitionArgs = SP->getType()->getTypeArray();
+
+      if (DeclArgs.size() && DefinitionArgs.size())
```

Comment:
Explanatory comment needed -- this is over-riding the declarations type with that of the definitions type? (If so, why?)